### DISCUSSION TITLE
Proposal Generic cast operator

### DISCUSSION THREAD

**Discussion by gafter** [dotnet/csharplang#8723] [Language Ideas]
@NN--- commented on [Thu Jun 16 2016](https://github.com/dotnet/roslyn/issues/12043)

I would like to have this code to be allowed.

``` cs
class A
{
  public static implicit operator<T> A<T>(T t) { return new A(); }
}

A a;
a = 1;
```

It is possible to do it in other languages such as [Nemerle](http://nemerle.org)

``` nemerle
class A
{
 public static @:[T](t : T) : A { A() } // "@:" in C# is implicit operator
} 

class Program
{
 public static Main() : void 
 {
  mutable a : A; // C#: A a;
  a = 1; 
 }
}
```

---

@benaadams commented on [Thu Jun 16 2016](https://github.com/dotnet/roslyn/issues/12043#issuecomment-226419828)

Wouldn't that need to be:

``` csharp
class A<T>
{
  public static implicit operator A<T>(T t) { return new A<T>(); }
}
```

---

@dsaf commented on [Thu Jun 16 2016](https://github.com/dotnet/roslyn/issues/12043#issuecomment-226420650)

@benaadams it's easier to see the difference with methods:

``` csharp
public IEnumerable Test<T>()
{
    return new object[] {};
}

public IEnumerable<T> Test<T>()
{
    return Enumerable.Empty<T>();
}
```

---

@HaloFour commented on [Thu Jun 16 2016](https://github.com/dotnet/roslyn/issues/12043#issuecomment-226461410)

The problem is that the syntax that you've described is already legal for a conversion operator.  As the generic type parameter is embedded in the type name which is reused as the operator method name syntactically you'd need another place to stick the generic type parameter in said operators.

---

@NN--- commented on [Thu Jun 16 2016](https://github.com/dotnet/roslyn/issues/12043#issuecomment-226471278)

@HaloFour Correct, didn't think about it.
It should be then something like:

``` cs
public static implicit operator<T> A<T>(T t) { return new A(); }
```

---

@HaloFour commented on [Thu Jun 16 2016](https://github.com/dotnet/roslyn/issues/12043#issuecomment-226477741)

@NN--- 

Or for your example:

``` cs
public static implicit operator<T> A(T t) { return new A(); }
```

Which is weird looking but there probably isn't a better candidate for locating a potential place to insert another generic type parameter.

I wonder how truly generic conversion operators would work given the limitations imposed on conversion operators to not allow conversions between base and derived types.  Although I guess the compiler allows some of those shenanigans today with conversion operators on generic classes.

---

@alrz commented on [Thu Jun 16 2016](https://github.com/dotnet/roslyn/issues/12043#issuecomment-226481769)

Using #11159,

``` cs
extension class Extensions<T> : A {
  public static explicit operator A(T t) => new A();
}
```

Not sure about implicit conversions though.

---

@NN--- commented on [Thu Jun 16 2016](https://github.com/dotnet/roslyn/issues/12043#issuecomment-226510187)

@HaloFour This what happens if your syntax requires types before names;)

**Comment by thomaslevesque** [comment]
This would be very useful. Basically, any time you need a conversion from a non-generic type `N` to a generic type `G<T>`. If you control `G<T>`, there's an easy workaround: just put the conversion from `N` in `G<T>`. But if you *don't* control `G<T>` (e.g. you want a conversion from `N` to `Action<T>`), you're stuck.

**Comment by Denis535** [comment]
I am surprised that I cannot do so:
```
public static implicit operator<TValue> MyKeyValueBase((string Key, TValue Value) pair) {
    return MyKeyValue<T>( pair.Key, pair.Value );
}
```

**Comment by CyrusNajmabadi** [comment]
@Denis535 that seems super unsafe.  Say you have:

```c#
interface ITuple { }
class MyTuple { }
class Derived : MyTuple, ITuple { }
// ...
ITuple t = new Derived();
MyTuple m = (MyTuple)t;
```

This would now call your implicit operator instead of the standard interface/class reference cast.

**Comment by znakeeye** [comment]
With the introduction of the [IParsable&lt;TSelf&gt;](https://learn.microsoft.com/en-us/dotnet/api/system.iparsable-1?view=net-7.0) interface in `.NET 7` the need for a generic cast operator actually got much bigger. The code below speaks for itself.

```c#
public readonly struct Variant
{
    private readonly string value;

    public Variant(string value) => this.value = value;

    public override string ToString() => value;

    public static implicit operator Variant(string s) => new(s);
    public static implicit operator string(Variant o) => o.value;

    // TODO: In future version of C#, replace with one and only one operator<T>
    // https://github.com/dotnet/csharplang/issues/534

#if HAS_GENERIC_CAST_OPERATOR
    // UPDATE: Calling Cast<T>() method instead of T.Parse()
    // public static implicit operator<T> T(Variant o) where T : IParsable<T> => T.Parse(this, null);
    public static implicit operator<T> T(Variant o) where T : IParsable<T> => o.Cast<T>();
#else
    public static implicit operator DateTime(Variant o) => o.Cast<DateTime>();
    public static implicit operator byte(Variant o) => o.Cast<byte>();
    public static implicit operator short(Variant o) => o.Cast<short>();
    public static implicit operator int(Variant o) => o.Cast<int>();
    public static implicit operator long(Variant o) => o.Cast<long>();
    // ...
    // ...
    // Keep 'em coming!
    public static implicit operator decimal(Variant o) => o.Cast<decimal>();
    // ...
#endif
    public TSelf Cast<TSelf>() where TSelf : IParsable<TSelf> => TSelf.Parse(this, null);
}
```

**Comment by bill-poole** [comment]
@znakeeye, I'm not sure I understand what the generic cast is doing in your example. Cast operators are `static` (as is yours), but it references `this`, which isn't possible in a `static` method. Furthermore, the `IParsable<T>.Parse` method accepts a `string` parameter, whereas `this` is of type `Variant`. Do you mean `T.Parse(o.value, null)`?

**Comment by bill-poole** [comment]
I think the value of and need for generic cast operators in C# has increased now that we have generic math in C#.

**Comment by znakeeye** [comment]
> @znakeeye, I'm not sure I understand what the generic cast is doing in your example. Cast operators are `static` (as is yours), but it references `this`, which isn't possible in a `static` method. Furthermore, the `IParsable<T>.Parse` method accepts a `string` parameter, whereas `this` is of type `Variant`. Do you mean `T.Parse(o.value, null)`?

I think my example clearly highlights the problem *and* the proposed solution. Without support for a generic cast operator, you have to write one operator for each type, which makes no sense now that we have a generic `T.Parse()` method.

I'm a bit surprised that this proposal hasn't gained more attention.

**Comment by bill-poole** [comment]
But your `static` cast operator has `this` in it, which is not permitted in a `static` method/operator. And `this` is not of type `string`, which is the type of the first parameter passed to `IParsable<T>.Parse` method.

**Comment by znakeeye** [comment]
Right. So maybe we should have written it like this instead?
```
public static implicit operator<T> T(Variant o) where T : IParsable<T> => o.Cast<T>();
```

**Comment by bill-poole** [comment]
That looks right to me.

**Comment by panagiotidisc** [comment]
What I find interesting is this proposal, along with generic/templated constructors, would allow us to write a poor man's covariant generic class (classes should be able to be covariant and contravariant like interfaces in my honest opinion)
```
public abstract class BaseClass
{
	public BaseClass() { }
}
public class DerivedClass : BaseClass
{
	public DerivedClass() { }
}
public abstract class Clazz
{
	protected delegate void FooDelegate();
	protected abstract FooDelegate fooDelegate { get; }
	public void Foo()
	{
		fooDelegate();
	}
}
public class CovariantGeneric<T1> : Clazz where T1 : BaseClass
{
	public T1 bar { get; }

	protected CovariantGeneric() { }
	
	//maybe CovariantGeneric<T1, T2> would be better
	private CovariantGeneric<T2>(CovariantGeneric<T2> other) where T2 : T1
	{
		bar = other.bar;
		fooDelegate = other.fooDelegate;
	}
	public static implicit operator<T2> CovariantGeneric<T1>(CovariantGeneric<T2> other) where T2 : T1
	{
		return new CovariantGeneric<T1>(other);
	}
}
```
